File: backend/src/services/cache.js
/**
 * Redis Caching Service
 * Provides caching for AI verification results and generated images
 */

import { createHash } from 'crypto';
import Redis from 'ioredis';

// Initialize Redis client
const REDIS_URL = process.env.HACKNYU_REDIS_URL || 'redis://localhost:6379';
const REDIS_ENABLED = process.env.HACKNYU_REDIS_ENABLED !== 'false';

let redisClient = null;
let isConnected = false;

if (REDIS_ENABLED) {
  try {
    redisClient = new Redis(REDIS_URL, {
      retryStrategy(times) {
        const delay = Math.min(times * 50, 2000);
        console.log(`[Redis] Retrying connection in ${delay}ms...`);
        return delay;
      },
      maxRetriesPerRequest: 3,
      enableReadyCheck: true,
    });

    redisClient.on('connect', () => {
      console.log('[Redis] Connected successfully');
      isConnected = true;
    });

    redisClient.on('error', (error) => {
      console.error('[Redis] Connection error:', error.message);
      isConnected = false;
    });

    redisClient.on('close', () => {
      console.log('[Redis] Connection closed');
      isConnected = false;
    });
  } catch (error) {
    console.error('[Redis] Initialization error:', error.message);
  }
} else {
  console.log('[Redis] Caching disabled');
}

/**
 * Generate a hash from data for use as cache key
 * @param {string|object} data - Data to hash
 * @returns {string} SHA-256 hash
 */
function generateHash(data) {
  const content = typeof data === 'string' ? data : JSON.stringify(data);
  return createHash('sha256').update(content).digest('hex');
}

/**
 * Generate cache key for verification results
 * @param {string} imageData - Base64 image data
 * @param {string} modelId - Model ID used for verification
 * @returns {string} Cache key
 */
function getVerificationCacheKey(imageData, modelId) {
  // Hash the image to create a unique key
  const imageHash = generateHash(imageData);
  return `verification:${modelId}:${imageHash}`;
}

/**
 * Generate cache key for generated images
 * @param {string} prompt - Image generation prompt
 * @param {string} modelId - Model ID used for generation
 * @returns {string} Cache key
 */
function getImageGenCacheKey(prompt, modelId) {
  const promptHash = generateHash(prompt);
  return `image_gen:${modelId}:${promptHash}`;
}

/**
 * Get verification result from cache
 * @param {string} imageData - Base64 image data
 * @param {string} modelId - Model ID
 * @returns {Promise<object|null>} Cached result or null
 */
export async function getCachedVerification(imageData, modelId) {
  if (!redisClient || !isConnected) {
    console.log('[Redis] Cache not available, skipping');
    return null;
  }

  try {
    const key = getVerificationCacheKey(imageData, modelId);
    const cached = await redisClient.get(key);

    if (cached) {
      console.log(`[Redis] Cache HIT for verification (model: ${modelId})`);
      const result = JSON.parse(cached);
      result._metadata = result._metadata || {};
      result._metadata.cacheHit = true;
      return result;
    }

    console.log(`[Redis] Cache MISS for verification (model: ${modelId})`);
    return null;
  } catch (error) {
    console.error('[Redis] Error getting cached verification:', error.message);
    return null;
  }
}

/**
 * Cache verification result
 * @param {string} imageData - Base64 image data
 * @param {string} modelId - Model ID
 * @param {object} result - Verification result to cache
 * @param {number} ttl - Time to live in seconds (default: 24 hours)
 * @returns {Promise<boolean>} Success status
 */
export async function cacheVerification(imageData, modelId, result, ttl = 86400) {
  if (!redisClient || !isConnected) {
    return false;
  }

  try {
    const key = getVerificationCacheKey(imageData, modelId);
    await redisClient.setex(key, ttl, JSON.stringify(result));
    console.log(`[Redis] Cached verification result (TTL: ${ttl}s)`);
    return true;
  } catch (error) {
    console.error('[Redis] Error caching verification:', error.message);
    return false;
  }
}

/**
 * Get generated image URL from cache
 * @param {string} prompt - Image generation prompt
 * @param {string} modelId - Model ID
 * @returns {Promise<string|null>} Cached image URL or null
 */
export async function getCachedImageUrl(prompt, modelId) {
  if (!redisClient || !isConnected) {
    return null;
  }

  try {
    const key = getImageGenCacheKey(prompt, modelId);
    const cached = await redisClient.get(key);

    if (cached) {
      console.log(`[Redis] Cache HIT for image generation (model: ${modelId})`);
      return cached;
    }

    console.log(`[Redis] Cache MISS for image generation (model: ${modelId})`);
    return null;
  } catch (error) {
    console.error('[Redis] Error getting cached image:', error.message);
    return null;
  }
}

/**
 * Cache generated image URL
 * @param {string} prompt - Image generation prompt
 * @param {string} modelId - Model ID
 * @param {string} imageUrl - Generated image URL
 * @param {number} ttl - Time to live in seconds (default: 7 days)
 * @returns {Promise<boolean>} Success status
 */
export async function cacheImageUrl(prompt, modelId, imageUrl, ttl = 604800) {
  if (!redisClient || !isConnected) {
    return false;
  }

  try {
    const key = getImageGenCacheKey(prompt, modelId);
    await redisClient.setex(key, ttl, imageUrl);
    console.log(`[Redis] Cached image URL (TTL: ${ttl}s)`);
    return true;
  } catch (error) {
    console.error('[Redis] Error caching image URL:', error.message);
    return false;
  }
}

/**
 * Invalidate cached verification result
 * @param {string} imageData - Base64 image data
 * @param {string} modelId - Model ID
 * @returns {Promise<boolean>} Success status
 */
export async function invalidateVerification(imageData, modelId) {
  if (!redisClient || !isConnected) {
    return false;
  }

  try {
    const key = getVerificationCacheKey(imageData, modelId);
    await redisClient.del(key);
    console.log('[Redis] Invalidated verification cache');
    return true;
  } catch (error) {
    console.error('[Redis] Error invalidating cache:', error.message);
    return false;
  }
}

/**
 * Get cache statistics
 * @returns {Promise<object>} Cache statistics
 */
export async function getCacheStats() {
  if (!redisClient || !isConnected) {
    return {
      enabled: false,
      connected: false,
    };
  }

  try {
    const info = await redisClient.info('stats');
    const dbSize = await redisClient.dbsize();

    // Parse info string
    const stats = {};
    info.split('\r\n').forEach(line => {
      const [key, value] = line.split(':');
      if (key && value) {
        stats[key] = value;
      }
    });

    return {
      enabled: true,
      connected: isConnected,
      totalKeys: dbSize,
      keyspaceHits: parseInt(stats.keyspace_hits) || 0,
      keyspaceMisses: parseInt(stats.keyspace_misses) || 0,
      hitRate: stats.keyspace_hits && stats.keyspace_misses
        ? (parseInt(stats.keyspace_hits) / (parseInt(stats.keyspace_hits) + parseInt(stats.keyspace_misses)) * 100).toFixed(2) + '%'
        : 'N/A',
    };
  } catch (error) {
    console.error('[Redis] Error getting stats:', error.message);
    return {
      enabled: true,
      connected: isConnected,
      error: error.message,
    };
  }
}

/**
 * Clear all cache entries (use with caution!)
 * @param {string} pattern - Pattern to match keys (optional, default: all)
 * @returns {Promise<number>} Number of keys deleted
 */
export async function clearCache(pattern = '*') {
  if (!redisClient || !isConnected) {
    return 0;
  }

  try {
    const keys = await redisClient.keys(pattern);
    if (keys.length === 0) {
      console.log('[Redis] No keys to delete');
      return 0;
    }

    const deleted = await redisClient.del(...keys);
    console.log(`[Redis] Deleted ${deleted} cache entries`);
    return deleted;
  } catch (error) {
    console.error('[Redis] Error clearing cache:', error.message);
    return 0;
  }
}

/**
 * Close Redis connection (for graceful shutdown)
 */
export async function closeRedisConnection() {
  if (redisClient) {
    try {
      await redisClient.quit();
      console.log('[Redis] Connection closed gracefully');
    } catch (error) {
      console.error('[Redis] Error closing connection:', error.message);
    }
  }
}

export default {
  getCachedVerification,
  cacheVerification,
  getCachedImageUrl,
  cacheImageUrl,
  invalidateVerification,
  getCacheStats,
  clearCache,
  closeRedisConnection,
};

File: backend-py/app/services/cache.py
"""Redis caching service (async port of Node `services/cache.js`).

Caches AI verification results (24h) and generated image URLs (7d). The client
is a lazily-initialised module-level singleton reused across warm Lambda
invocations.

Caching is strictly optional: when `HACKNYU_REDIS_ENABLED` is false, or Redis is
unreachable, every operation degrades to a no-op that warns and continues — it
must never raise into the request path.
"""

from __future__ import annotations

import hashlib
import json
import logging
from typing import Any

import redis.asyncio as redis

from app.config.settings import get_settings

logger = logging.getLogger(__name__)

# TTLs (seconds) — match the Express service.
VERIFICATION_TTL = 86_400  # 24 hours
IMAGE_URL_TTL = 604_800  # 7 days

# Module-level singleton, lazily initialised and reused warm.
_client: redis.Redis | None = None
_init_attempted = False


def _get_client() -> redis.Redis | None:
    """Return the shared Redis client, or None if caching is disabled/unavailable.

    Initialisation is lazy and attempted only once. Construction failures are
    swallowed (warn-and-continue); connection failures surface later on the
    first command and are handled there.
    """
    global _client, _init_attempted

    if _init_attempted:
        return _client

    _init_attempted = True
    settings = get_settings()

    if not settings.hacknyu_redis_enabled:
        logger.info("[Redis] Caching disabled")
        return None

    try:
        _client = redis.Redis.from_url(
            settings.hacknyu_redis_url,
            decode_responses=True,
            socket_connect_timeout=2,
            socket_timeout=2,
            retry_on_timeout=False,
        )
    except Exception as error:  # noqa: BLE001 — never fail init
        logger.error("[Redis] Initialization error: %s", error)
        _client = None

    return _client


def reset_cache_client() -> None:
    """Reset the module-level singleton. Test-only helper."""
    global _client, _init_attempted
    _client = None
    _init_attempted = False


def _generate_hash(data: Any) -> str:
    """SHA-256 hex digest of `data` (stringified if not already a string)."""
    content = data if isinstance(data, str) else json.dumps(data)
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


def _verification_cache_key(image_data: str, model_id: str) -> str:
    return f"verification:{model_id}:{_generate_hash(image_data)}"


def _image_gen_cache_key(prompt: str, model_id: str) -> str:
    return f"image_gen:{model_id}:{_generate_hash(prompt)}"


async def get_cached_verification(image_data: str, model_id: str) -> dict[str, Any] | None:
    """Return a cached verification result (with `_metadata.cacheHit=True`) or None."""
    client = _get_client()
    if client is None:
        logger.debug("[Redis] Cache not available, skipping")
        return None

    try:
        key = _verification_cache_key(image_data, model_id)
        cached = await client.get(key)
        if cached:
            logger.info("[Redis] Cache HIT for verification (model: %s)", model_id)
            result = json.loads(cached)
            result["_metadata"] = result.get("_metadata") or {}
            result["_metadata"]["cacheHit"] = True
            return result
        logger.info("[Redis] Cache MISS for verification (model: %s)", model_id)
        return None
    except Exception as error:  # noqa: BLE001 — never fail the request
        logger.error("[Redis] Error getting cached verification: %s", error)
        return None


async def cache_verification(
    image_data: str,
    model_id: str,
    result: dict[str, Any],
    ttl: int = VERIFICATION_TTL,
) -> bool:
    """Store a verification result. Returns True on success, False otherwise."""
    client = _get_client()
    if client is None:
        return False

    try:
        key = _verification_cache_key(image_data, model_id)
        await client.set(key, json.dumps(result), ex=ttl)
        logger.info("[Redis] Cached verification result (TTL: %ss)", ttl)
        return True
    except Exception as error:  # noqa: BLE001 — never fail the request
        logger.error("[Redis] Error caching verification: %s", error)
        return False


async def get_cached_image_url(prompt: str, model_id: str) -> str | None:
    """Return a cached generated-image URL or None."""
    client = _get_client()
    if client is None:
        return None

    try:
        key = _image_gen_cache_key(prompt, model_id)
        cached = await client.get(key)
        if cached:
            logger.info("[Redis] Cache HIT for image generation (model: %s)", model_id)
            return cached
        logger.info("[Redis] Cache MISS for image generation (model: %s)", model_id)
        return None
    except Exception as error:  # noqa: BLE001 — never fail the request
        logger.error("[Redis] Error getting cached image: %s", error)
        return None


async def cache_image_url(
    prompt: str,
    model_id: str,
    image_url: str,
    ttl: int = IMAGE_URL_TTL,
) -> bool:
    """Store a generated-image URL. Returns True on success, False otherwise."""
    client = _get_client()
    if client is None:
        return False

    try:
        key = _image_gen_cache_key(prompt, model_id)
        await client.set(key, image_url, ex=ttl)
        logger.info("[Redis] Cached image URL (TTL: %ss)", ttl)
        return True
    except Exception as error:  # noqa: BLE001 — never fail the request
        logger.error("[Redis] Error caching image URL: %s", error)
        return False


async def invalidate_verification(image_data: str, model_id: str) -> bool:
    """Delete a cached verification result. Returns True on success."""
    client = _get_client()
    if client is None:
        return False

    try:
        key = _verification_cache_key(image_data, model_id)
        await client.delete(key)
        logger.info("[Redis] Invalidated verification cache")
        return True
    except Exception as error:  # noqa: BLE001 — never fail the request
        logger.error("[Redis] Error invalidating cache: %s", error)
        return False
